aloha-io-avro/src/main/scala/com/eharmony/aloha/audit/impl/avro/AvroGenericRecordAuditor.scala

```diff
@@ -0,0 +1,215 @@
+import com.eharmony.aloha.reflect.{RefInfo, RefInfoOps}
+import org.apache.avro.Schema
+import org.apache.avro.Schema.Type.RECORD
+import org.apache.avro.generic.{GenericData, GenericRecord}
+import org.apache.commons.io.IOUtils
+import java.{util => ju}
+import scala.collection.JavaConversions.{seqAsJavaList, collectionAsScalaIterable}
+
+/**
+  * An auditor for encoding information in a tree structure as an Avro `org.apache.avro.generic.GenericRecord`.
+  *
+  * See the resources in [[com.eharmony.aloha.audit.impl.avro]] for the Avro schema.
+  *
+  * Created by ryan on 2/26/17.
+  *
+  * @tparam N the natural output type of a model whose data is to be audited.
+  */
+sealed abstract class AvroGenericRecordAuditor[N]
+  extends MorphableAuditor[GenericRecord, N, GenericRecord]
+     with Serializable
+
+object AvroGenericRecordAuditor extends Serializable {
+  /**
+    * '''NOTE''': The class loader is used to locate the Avro schema specification rather than
+    * Apache VFS 1 or 2.  This is done because issues can arise when users initialize both
+    * VFS 1 and 2.  Since the choice of auditor shouldn't affect the user's choice or VFS
+    * version elsewhere in Aloha, we don't use VFS.
+    */
+  private[this] val SchemaResourceLocation = "com/eharmony/aloha/audit/impl/avro/generic_record_auditor.avsc"
+
+  // TODO: Should the schemas be inside AvroGenericRecordAuditor instead of in the companion object?
+
+  /**
+    * `org.apache.avro.Schema` doesn't extend Serializable so this needs to be transient.
+    */
+  @transient private[this] lazy val AuditorSchema: Schema = {
+    val parser = new Schema.Parser()
+    val is = getClass.getClassLoader.getResourceAsStream(SchemaResourceLocation)
+    try {
+      parser.parse(is)
+    }
+    finally {
+      IOUtils.closeQuietly(is)
+    }
+  }
+
+  /**
+    * `org.apache.avro.Schema` doesn't extend Serializable so this needs to be transient.
+    */
+  @transient private[this] lazy val ModelIdSchema: Schema = {
+    // the `.get` at the end blows up (on purpose).
+    collectionAsScalaIterable(AuditorSchema.getField("model").schema().getTypes).toSeq.find(s => s.getType == RECORD).get
+  }
+
+  // Force the retrieval of the lazy vals at creation time.  They are only lazy to avoid serializing.
+  require(AuditorSchema != null && ModelIdSchema != null)
+
+  def apply[N](implicit ri: RefInfo[N]): Option[AvroGenericRecordAuditor[N]] = {
+    ri match {
+      case RefInfo.Boolean => opt(IdentityAuditor[Boolean]())
+      case RefInfo.Byte => opt(IdentitySubTypeAuditor[Byte, Int])
+      case RefInfo.Short => opt(IdentitySubTypeAuditor[Short, Int])
+      case RefInfo.Int => opt(IdentityAuditor[Int]())
+      case RefInfo.Long  => opt(IdentityAuditor[Long]())
+      case RefInfo.Float => opt(IdentityAuditor[Float]())
+      case RefInfo.Double => opt(IdentityAuditor[Double]())
+
+      // TODO: Should a String really be inserted to the record or a CharSequence like a avro.util.Utf8?
```

Comment:
I'll try to get this done.  Don't have much experience but I'll try to knock it out.